### ISSUE TITLE
Perform perturbation screening not reproducible?

### ISSUE THREAD

**Issue by Melody-cell** [theislab/regvelo#19]
Hi, thanks for the great tool. I have met a problem here:

when I run the example data given in the tutorial, i can't reproduce this result, my result is as follow:

![Image](https://github.com/user-attachments/assets/269b81de-d951-4ab5-b2f6-73b12098e66d)

it's not the same as given in the tutorial.

Is this normal and why?

Best,
Melody

**Comment by WWXkenmo** [comment]
Hey Melody

Could you please let me know your scvi version at first? and all the code that you have  used to run tutorial?

Best Regards
Ken

**Comment by WWXkenmo** [comment]
It might relate to this issue (https://github.com/theislab/regvelo/issues/15), if your scvi-tools is not in 1.2.0 version, please set accordingly

Best,
Ken

**Comment by WWXkenmo** [comment]
Hey Melody @Melody-cell 

Did you have checked if issue still there when you change scvi-tools version?

If still not work, I assume there has other internal conflicts between different packages, please used newly submitted yaml file in `reproduce_env` folder, and create new conda environment and have try

`conda env create -f regvelo.yaml -n regvelo-py310`

I have asked my colleague to reproduce and so far we didn't find issue, but please let me know if it's still not working for you, it's very important for me

Many thanks!

**Comment by Melody-cell** [comment]
Hey Ken, this is my scvi version:

![Image](https://github.com/user-attachments/assets/c6523cae-efdc-42ff-8ea4-4bf4de11b9e1)

And this is my code:

import numpy as np
import pandas as pd
import scanpy as sc
import scvelo as scv
import torch

import cellrank as cr
import scvi

import matplotlib.pyplot as plt
import seaborn as sns

from velovi import preprocess_data
import regvelo
from regvelo import REGVELOVI,sanity_check,set_prior_grn,abundance_test,TFscreening

pip install mplscience
`cause conda install failed, so i use pip`
import mplscience

scvi.settings.seed = 0 

import anndata as ad
adata = ad.read_h5ad("/home/wangmengj/workspace/project04_scenicplus/a07_regvelo/01_example_tutorial/data/adata_zebrafish_preprocessed.h5ad")
prior_net = pd.read_csv("/home/wangmengj/workspace/project04_scenicplus/a07_regvelo/01_example_tutorial/data/prior_GRN.csv", index_col = 0)
TF_list = adata.var_names[adata.var["is_tf"]].tolist()

sc.pp.neighbors(adata, n_neighbors=30, n_pcs=50)
scv.pp.moments(adata)
adata.layers["spliced"] 
sc.pl.umap(adata, color='cell_type', palette=sc.pl.palettes.vega_20)
adata = preprocess_data(adata)
adata = set_prior_grn(adata, prior_net)
TF_list = set(TF_list).intersection(adata.var_names)
TF_list = list(TF_list)
print("final number of TF: "+str(len(TF_list)))
W = adata.uns["skeleton"].copy()
W = torch.tensor(np.array(W))
W
sparsity = W.sum() / ((W.sum(1) != 0).sum() * W.shape[0])
print("network sparsity: "+str(np.array(sparsity)))


As for Training RegVelo model, i ran it in shell script.

As for Perform perturbation screening: 
ct_indices = {}
for ct in terminal_states:
    indices = adata_target.obs["term_states_fwd"][adata_target.obs["term_states_fwd"] == ct].index.tolist()
    ct_indices[ct] = indices


fate_prob_perturb = []
cand_list = ["nr2f5","sox9b","twist1a","twist1b","sox10","mitfa","tfec","tfap2b"]
for TF in cand_list:
    model = "/home/wangmengj/workspace/project04_scenicplus/a07_regvelo/01_example_tutorial/outs/rgv_model"
    adata_target_perturb,reg_vae_perturb =  regvelo.in_silico_block_simulation(model,adata,TF,cutoff = 0)
    ### predict the terminal state dfference
    n_states = 7
    vk = cr.kernels.VelocityKernel(adata_target_perturb).compute_transition_matrix()
    estimator = cr.estimators.GPCCA(vk)
    ## evaluate the fate prob on original space
    estimator.compute_macrostates(n_states=n_states,cluster_key="cell_type")
    
    
    estimator.set_terminal_states(ct_indices)
    estimator.compute_fate_probabilities()
    ## visualize coefficient
    cond1_df = pd.DataFrame(adata_target_perturb.obsm["lineages_fwd"], columns=adata_target_perturb.obsm["lineages_fwd"].names.tolist())
    cond1_df = cond1_df.loc[:,terminal_states]
    
    fate_prob_perturb.append(cond1_df)

In this step, it appears an error:

![Image](https://github.com/user-attachments/assets/7f495c41-e0cf-43d3-a404-372fe934a7c0)

So i changed one parameter: 

![Image](https://github.com/user-attachments/assets/66b8165c-ffe4-4d03-a9bb-3824b9cf0215)


I will try this soon `conda env create -f regvelo.yaml -n regvelo-py310`, I'll let you know if it's still not working for me.
Thanks a lot!

**Comment by WWXkenmo** [comment]
Hey Melody,

Thanks for let me know, I think it mainly due to some unexpected package issue, but I don't know where could be.

I think another possible way is to force reinstall but rather use the package in cache of pip, in which you can

`pip install git+https://github.com/theislab/regvelo.git@main --no-cache-dir --force-reinstall`

Let me know if it also work for you or not

Best Regards,
Ken

**Comment by Melody-cell** [comment]
Hi Ken,
I have tried this: `conda env create -f regvelo.yaml -n regvelo-py310`, it failed:

[reproduce_env.log](https://github.com/user-attachments/files/19459326/reproduce_env.log)

Then i tried this: `pip install git+https://github.com/theislab/regvelo.git@main --no-cache-dir --force-reinstall`, 
but when `import mplscience` it appears an error, i guess it may be this package issue?
![Image](https://github.com/user-attachments/assets/a6a7126c-7431-4e4b-869d-4574bf55b600)

**Comment by WWXkenmo** [comment]
OK great, for mplscience is actually a package for plotting figure, just install it

`pip install mplscience`

let me know how results look like

Best,
Ken

**Comment by Melody-cell** [comment]
Hi, @WWXkenmo ,
it appeared this error as the last time:

![Image](https://github.com/user-attachments/assets/989ca18f-6eed-4e53-82db-ec5e821df96a)

And I found it stopped at Epoch 1244/1500:  83%, different from it in the tutorial: Epoch 1410/1500:  94%.

![Image](https://github.com/user-attachments/assets/a5dd628d-0552-4d92-b674-a5c1c4cbce99)

**Comment by WWXkenmo** [comment]
Hi @Melody-cell 

This error is related to the PyTorch version. After version 2.6, PyTorch changed the default arguments for `weight_only`.

I'll explicitly set the PyTorch version requirement to <2.6 in the current package. For now, you can resolve this issue by downgrading PyTorch to version 2.5.1.

Regarding the early stopping behavior, it depends a lot of thing, for example, Initialization values, the device you're using (CPU/GPU), and even the PyTorch version itself.

The behavior doesn't need to exactly match my tutorial. For example In the tutorial, I used an A100 GPU, but your setup might be using a CPU (since each iteration is taking much longer).

If you have a GPU available, could you confirm whether CUDA is properly configured?

Let me know if you need further help!

All the best,
Ken

**Comment by WWXkenmo** [comment]
Hey Melody,

I just redo installation again, and everything is fine for me.

I hope it will work with you, just don't hesitate to ask your issue, thanks a lot for letting me know on these.

Best,
Ken

**Comment by Melody-cell** [comment]
Hi, I tried it again, the result is as follow:

![Image](https://github.com/user-attachments/assets/860d3f48-ef39-4e15-a73e-5fbcc631c6c1)

but still slightly different from the tutorial? see the red box below；

![Image](https://github.com/user-attachments/assets/86a7fb9b-fdbf-42bc-bfc7-ca6609194d5c)

**Comment by WWXkenmo** [comment]
Hi,

Thanks a lot! this slightly difference is not problem: it doesn't influence your conclusion, actually twist1b is the drivers relate to facial mesenchymal cells, which you also have show this behavior, after knock out twist1b, the facial mesenchymal cell is depleted.

DL model doesn't have deterministic solution like linear model, it always have randomness which can be caused by different initializations, device type and so on. 

I hope it can address your concerns.

Best Regards,
Ken

**Comment by Melody-cell** [comment]
OK, i got it, thanks a lot !


Best,
Melody
